### PR TITLE
Remove usage of old safe math and unused intsafe.h include

### LINKED ISSUE #4153: Remove usages of old safe math
There are approximately 79 uses of old safe math. (a.k.a. the `LongAdd`, `ShortAdd`, and friends from `intsafe.h`).

This task represents removing them and transitioning them to an appropriate safe/saturating math function from the chromium safe math library introduced in #4144.

After removing usages of the old safe math functions, also remove the `intsafe.h` header.

Files with old safe math as of this writing:
- TerminalControl: TSFInputControl.cpp
- TerminalCore: TerminalDispatch.cpp
- TerminalCore: TerminalSelection.cpp
- Host: directio.cpp
- RendererGdi: invalidate.cpp
- RendererGdi: math.cpp
- RendererGdi: paint.cpp
- RendererVt: paint.cpp
- TerminalAdapter: adaptDispatch.cpp
- Types: viewport.cpp
- Types: WindowUiaProviderBase.cpp

### PR DESCRIPTION
**Pull request by skyline75489**
## References

#4144 introduced the chromium safe math library.

## PR Checklist
* [X] Closes #4153
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

### CODE REVIEW

File: src/interactivity/win32/precomp.h
Comment (skyline75489) [inline src/interactivity/win32/precomp.h:7]: I remember that interactivity should not be changed in this repo. But I removed the `<intsafe.h>` in `../../host/precomp.h`  so I had to add it back here to make the project compile.

**Comment by j4james** [comment]
From the discussion we had in issue #4013, I got the impression we'd be getting cleaner, more readable code from the new safe math library, so I've got to say these changes are a little disappointing. What exactly are the benefits we're getting here? 

And it looks like the new lib is introducing some behavioural changes too - the overflow tests look slightly different to me, and the error codes returned are different. Is this intentional? And the one behavioural change I was expecting was greater use of clamping instead of failing on overflow, but that doesn't seem to have happened.

If all we're doing is replacing one unreadable library with a slightly more unreadable library (and risking introducing new bugs in the process) is this really worth the effort?

**Review by zadjii-msft — APPROVED** [review]
This all seems sensible to me, thanks for cleaning this up!

**Comment by zadjii-msft** [comment]
@msftbot make sure @miniksa signs off on this one

File: src/host/directio.cpp
Comment (j4james) [inline src/host/directio.cpp:998]: Shouldn't all of these PtrdiffT calculations be using `ptrdiff_t` for the type? I don't think it's the same as `LONGLONG` in a 32bit build.

File: src/types/viewport.cpp
Comment (j4james) [inline src/types/viewport.cpp:854]: Shouldn't these `E_ABORT` errors be `INTSAFE_E_ARITHMETIC_OVERFLOW` if we want to remain compatible with the original code? It's probably not a big deal, but it's theoretically possible someone could be checking for specific error returns in some situation, especially if the return code is exposed in a public API somewhere.

Comment (j4james) [reply src/types/viewport.cpp:854]: Although I suppose if we want to get rid of the _intsafe.h_ dependency then I guess that's not an option, unless you want to redefine the constant, or hardcode the error number. I don't know.

**Comment by Jspooner09** [comment]
Have computer virus


Jimmy R. 

CONFIDENTIALITY: The information contained in this transmission may contain privileged and confidential information. It is intended only for the use of the person(s) named above. If you are not the intended recipient, you are hereby notified that any review, dissemination, distribution or duplication of this communication, and the information contained in it, is strictly prohibited. If you are not the intended recipient, please contact the sender and immediately destroy all copies of the original message.



On Thursday, December 3, 2020, 11:25 AM, msftbot[bot] <notifications@github.com> wrote:




Hello @zadjii-msft!

Because you've given me some instructions on how to help merge this pull request, I'll be modifying my merge approach. Here's how I understand your requirements for merging this pull request:
   
   - I'll only merge this pull request if it's approved by @miniksa

If this doesn't seem right to you, you can tell me to cancel these instructions and use the auto-merge policy that has been configured for this repository. Try telling me "forget everything I just told you".

File: src/types/viewport.cpp
Comment (miniksa) [reply src/types/viewport.cpp:854]: Hum. You're right that it probably should maintain the error code in case anyone's checking specifically for it.... but I also highly doubt that anyone is given that I probably introduced it in the last few years and most things I've ever found are only checking overall success/failure from any of our API returns (if they even check at all and don't just steamroll anyway).

If we're worried about it, I don't mind redefining that constant without the header if the goal is to remove the header. But also, I don't super mind if it just becomes a different error code either...

File: src/cascadia/TerminalCore/TerminalDispatch.cpp
Comment (miniksa) [inline src/cascadia/TerminalCore/TerminalDispatch.cpp:43]: @j4james had mentioned in the linked issue that we should probably replace these type conversions too to use the saturating math at the same time. If these aren't coming in from `intsafe.h`, where are they coming from? We should probably either do these in a batch with this PR or ensure we do a follow up to get them too for consistency.

**Comment by skyline75489** [comment]
@j4james 

> I got the impression we'd be getting cleaner, more readable code from the new safe math library, so I've got to say these changes are a little disappointing.

I feel about the same. I can’t say I fully endorse  this PR in terms of readability. But I still think that chromium safe math is better in terms of extensibility and modern code style. 

Well, “normal” cpp code does not use HRESULT as its return value and HRESULT is exactly what THROW wants. Maybe we just can not have the best of the both worlds, huh. 

I bet @miniksa can think of better reasons than me.

**Comment by j4james** [comment]
> Well, “normal” cpp code does not use HRESULT as its return value and HRESULT is exactly what THROW wants. Maybe we just can not have the best of the both worlds, huh.

Well what I was hoping would happen is we'd either be using clamped math, in which case there wouldn't be any error return value, or we'd be using a form of checked math that threw an exception, in which case the error handling could be dealt with neatly in one place (instead of every single expression having an error check on it).

And once we're not doing error checking on every operation, we're in a much better position to be using regular math operators with safe math types, which I think should be much more readable than all these `base::CheckXXX` calls. Maybe that's asking a bit much for the first PR, but hopefully we'll be able to get there eventually.

File: src/types/viewport.cpp
Comment (j4james) [reply src/types/viewport.cpp:854]: Yeah OK. The more I think about it, the more I think it's probably not a big deal, especially if this code was only a fairly recent addition.

File: src/cascadia/TerminalCore/TerminalDispatch.cpp
Comment (skyline75489) [reply src/cascadia/TerminalCore/TerminalDispatch.cpp:43]: Dang these aren't coming directly from `intsafe.h` but by including some win32 headers which indirectly include `intsafe.h`, these macros are still available. For example wil has it in https://github.com/microsoft/wil/blob/master/include/wil/safecast.h

This makes me feel that this PR isn't really doing much..

Comment (miniksa) [reply src/cascadia/TerminalCore/TerminalDispatch.cpp:43]: Hm. Darn ok.

**Comment by miniksa** [comment]
> > Well, “normal” cpp code does not use HRESULT as its return value and HRESULT is exactly what THROW wants. Maybe we just can not have the best of the both worlds, huh.
> 
> Well what I was hoping would happen is we'd either be using clamped math, in which case there wouldn't be any error return value, or we'd be using a form of checked math that threw an exception, in which case the error handling could be dealt with neatly in one place (instead of every single expression having an error check on it).
> 
> And once we're not doing error checking on every operation, we're in a much better position to be using regular math operators with safe math types, which I think should be much more readable than all these `base::CheckXXX` calls. Maybe that's asking a bit much for the first PR, but hopefully we'll be able to get there eventually.

That is true. The `base::CheckXXX` calls are really a kludge to deal with the existing workflow. The question is... do we take the baby step of getting everything to the chromium one first, even if it's ugly, then make a follow on to convert the ugly ones into a less ugly pattern? Or do we skip the first migration and just go straight for the second?

I know I'd find it mentally easier to do two more straight forward migrations than to have one more complicated one. But simultaneously, I can't see the core team getting to this any time soon... so I'm willing to concede whichever one is easier for the ones who are ambitious enough to do this work. If that's going straight to throwing-style calls with the bare operators on the safe math objects... (and of course catching the exception and routing it back to an HRESULT/NTSTATUS for the API call when necessary) that's fine with me.

> @j4james
> 
> > I got the impression we'd be getting cleaner, more readable code from the new safe math library, so I've got to say these changes are a little disappointing.
> 
> I feel about the same. I can’t say I fully endorse this PR in terms of readability. But I still think that chromium safe math is better in terms of extensibility and modern code style.
> 
> Well, “normal” cpp code does not use HRESULT as its return value and HRESULT is exactly what THROW wants. Maybe we just can not have the best of the both worlds, huh.
> 
> I bet @miniksa can think of better reasons than me.

This has just evolved from old Windows-style C code. Thus the return code focus and no real exceptions until recent-ish additions or modifications. That's why everything is so mixed. But we're ready and willing to adopt exceptions now as long as the console  API functionality is maintained to the perspective of clients (within reason). Thus if you want to close this issue/PR as "dumb/pointless/outdated" and transform the work into something that is overall prettier and more faithfully represents C++ and utilizes the Chromium math libraries to their potential... I'm cool with it.

**Comment by skyline75489** [comment]
After some consideration I'm still not convinced that this is good enough.

I'll close it for now. Thanks everyone for the explanation and comments.

**Pull request closed without merging**
